docs/how-to/hip_cpp_language_extensions.rst

```diff
@@ -0,0 +1,872 @@
+
+.. note::
+
+  ``warpSize`` should not be assumed to be a specific value in portable HIP
+  applications. NVIDIA devices return 32 for this variable; AMD devices return
+  64 for gfx9 and 32 for gfx10 and above. While code that assumes a ``warpSize``
+  of 32 can run on devices with a ``warpSize`` of 64, it only utilizes half of
+  the the compute resources.
+
+********************************************************************************
+Vector types
+********************************************************************************
+
+These types are not automatically provided by the compiler. The
+``hip_vector_types.h`` header, which is also included by ``hip_runtime.h`` has
+to be included to use these types.
+
+Fundamental vector types
+================================================================================
+
+Fundamental vector types derive from the `fundamental C++ integral and
+floating-point types <https://en.cppreference.com/w/cpp/language/types>`_. These
+types are defined in ``hip_vector_types.h``, which is included by
+``hip_runtime.h``.
+
+All vector types can be created with ``1``, ``2``, ``3`` or ``4`` elements, the
+corresponding type is ``<fundamental_type>i``, where ``i`` is the number of
+elements.
+
+All vector types support a constructor function of the form
+``make_<type_name>()``. For example,
+``float3 make_float3(float x, float y, float z)`` creates a vector of type
+``float3`` with value ``(x,y,z)``.
+The elements of the vectors can be accessed using their members ``x``, ``y``,
+``z``, and ``w``. 
+
+.. code-block:: cpp
+
+  double2 d2_vec = make_double2(2.0, 4.0);
+  double first_elem = d2_vec.x;
+
+HIP supports vectors created from the following fundamental types:
+
+.. list-table::
+  *
+    - :cspan:`1` Integral Types
+  *
+    - ``char``
+    - ``uchar``
+  *
+    - ``short``
+    - ``ushort``
+  *
+    - ``int``
+    - ``uint``
+  *
+    - ``long``
+    - ``ulong``
+  *
+    - ``longlong``
+    - ``ulonglong``
+  *
+    - :cspan:`1` Floating-Point Types
+  *
+    - :cspan:`1` ``float``
+  *
+    - :cspan:`1` ``double``
+
+.. _dim3:
+
+dim3
+================================================================================
+
+``dim3`` is a special three-dimensional unsigned integer vector type that is
+commonly used to specify grid and group dimensions for kernel launch
+configurations.
+
+Its constructor accepts up to three arguments. The unspecified dimensions are
+initialized to 1.
+
+********************************************************************************
+Built-in device functions
+********************************************************************************
+
+.. _memory_fence_instructions:
+
+Memory fence instructions
+================================================================================
+
+HIP supports ``__threadfence()``, ``__threadfence_block()`` and
+``__threadfence_system()``.
+
+On AMD devices, ``__threadfence_system()``, has restrictions and therefore needs
```

Comment:
This feels like it is referring to some implicit knowledge the user must have relative to threadfence_system? What are the "restrictions"? And is it appropriate to call it a workaround in the docs? Seems like a workaround is good for an issue or bug report, but not for user docs. Perhaps "requires the following process"?

Reply:
requires the following process?

Reply:
I'll ask around what the current status on this feature is, maybe we can get rid of this section altogether, but I'll postpone it for now